### PR TITLE
Autograd accum fix

### PR DESCRIPTION
**Pull request by linden-li**
This PR is a first attempt at fixing a synchronization bug with automatic gradient accumulation. There is an issue where some ranks perform their micro-batching loops and block when reducing their gradients while another rank has OOM'd. The fix right now tries to call a single backward pass in `_train_microbatch` in every iteration. If there is a failure in any of the ranks, it raises a `CUDA out of memory` issue, which is caught in the original logic.

### CODE REVIEW

File: composer/trainer/trainer.py [hunk 1]
```diff
@@ -1717,8 +1713,21 @@ def _train_microbatches(self, microbatches: Sequence[Batch], ddp_sync: bool = Tr
...
+                    if self.adaptive_gradient_accumulation and _is_cuda_oom(e):
+                        fail = 1
+                    else:
+                        raise
+
+                fail = self._device.tensor_to_device(torch.tensor([fail], dtype=torch.int8))
+                dist.all_reduce(fail, reduce_operation='SUM')
+
+                if fail.item() > 0:
+                    raise RuntimeError('CUDA out of memory')
```
Comment (ravi-mosaicml) [inline composer/trainer/trainer.py:1730]: Would it work to call the handle_cuda_oom here?

**Review by ravi-mosaicml — COMMENTED** [review]
Can you add a test case that would cause a hang before, and then that passes now with this fix? For the test setup, I think it could be done through an algorithm / callback that (massively) increases the batch size on rank 1 while keeping it the same on rank 0.

If it can wait, going to defer to @mvpatel2000 for approval. Otherwise, let's meet and reason through the distributed logic together to make sure we fix it once and for all.

**Comment by mvpatel2000** [comment]
Moving the discussion for the distributed logic offline -- it seems a little complicated

**Comment by mvpatel2000** [comment]
Closing this PR -- we discussed offline about a couple subtleties this doesn't address. Looks like this is pretty tricky, so I'll throw up a design doc to discuss a few alternatives

**Pull request closed without merging**